### ISSUE TITLE
macOS Big Sur real versioning hidden from pre-Big Sur binaries

### ISSUE THREAD

**Issue by mikdusan** [ziglang/zig#7569]
macOS (formerly branded Mac OS X) has been versioned fairly consistently since the initial `10.0` release about 20 years ago. Since then the major release number `10` has been used up until the release of Big Sur which bumped major to `11`.

Apple no doubt recognized that many existing applications need to detect macOS version at configure/build/runtime and decided to put in place a backwards compatible versioning scheme that is activated by two different means:

1. executables built against 10.15 SDK and earlier
2. executables launched with environ `SYSTEM_VERSION_COMPAT=1`

If either of these is true then the OS/kernel make it very difficult to obtain real versioning for Big Sur. In other words, Apple made it super easy for backwards version compatibility for ignorant executables, but made it nigh impossible for knowledgeable executables to detect Big Sur versioning.

But how and why is this important to Zig? Here's the problem statement:

**Zig posts release and CI binaries that are built on pre-Big Sur hosts and out of the box these binaries will not work on Big Sur as expected. Some of the issues have been worked around, but ultimately, host and (default) target information will be incorrect because the host will be detected as `10.16`.**

#### use sysctl MIB `kern.osproductversion`
Returns `10.16` inside a backwards compatibility context.

#### use sysctl MIB `kern.osversion` instead
We have a "temporary" fallback measure in place to convert macOS build strings to product versions and up to and including `10.15` it was formulaic and robust. By formulaic I mean that it would work with future versions. Unfortunately Big Sur completely breaks any formulaic pattern and build strings essentially are *monotonically increasing*. For example, we cannot predict when the first "letter" in the string will indicate a new minor or patch or both minor+patch change. As of this writing here are Big Sur known builds:

|Version|Build|
|-|-|
|11.0|20A2411|
|11.0.1|20B29|
|11.0.1|20B50|
|11.1|20C69|
|11.2 beta|20D5029f|

#### detect sysctl MIB `kern.osproductversioncompat`
It is possible to detect a **new** MIB that indicates current process is in backwards compatibility context. While this might be useful to emit a warning to find a better binary or some such, that's about as far as it goes. We still cannot determine Big Sur real version.

#### setenv `SYSTEM_VERSION_COMPAT=0` in zig compiler main
Does not work in the current process; it must be set in parent process of the executable.

#### have the user setenv `SYSTEM_VERSION_COMPAT=0` in their shell
This reeks. We'd like Zig binaries to just work out-of-the-box as much as possible.

#### read/parse file `/System/Library/CoreServices/SystemVersion.plist`
This was intended to be the go-to long term solution (see #5125) and is no longer viable. The file contents indicate `10.16` when in a backwards compatibility context.

#### parse output from `sw_vers`
Returns `10.16` inside a backwards compatibility context.

## Solution: a dotfile symlink
I did a deep search on a `11.1` system and found the following `.plist` files:

```
▸ ls -al /System/Library/CoreServices | grep ' SystemVersion.*plist'
lrwxr-xr-x   1 root wheel      19 Jan  1  2020 .SystemVersionPlatform.plist -> SystemVersion.plist
-r--r--r--   1 root wheel     524 Jan  1  2020 SystemVersion.plist
-r--r--r--   1 root wheel     529 Jan  1  2020 SystemVersionCompat.plist
```

It turns out that even though the symlink points to `SystemVersion.plist` reading/parsing file by that symlink path bypasses backwards compatibility context and always returns Big Sur real version info.

Thus our logic can have a single branch with something as straightforward as:

```
// `10.16` is only returned when in backwards compatible context
if sysctl MIB `kern.osproductversion` == `10.16` {
    parse_via_dotfile_simlink_path
}
```

**Comment by daurnimator** [comment]
Where/why do we care that the OSX release is big sur? and why can't we treat 10.6 the same as 11.1?

**Comment by mikdusan** [comment]
> Where/why do we care that the OSX release is big sur? and why can't we treat 10.6 the same as 11.1?

because many decisions such as clang or lld command line options are set according to what is detected. Using `10.16` here is meaningless; but using `11.1.0` is appropriate.

```
ld64.lld -error-limit 0 -demangle -dynamic -arch x86_64 -macosx_version_min 10.16.0 -sdk_version 10.16.0 -pie -o rf zig-cache/o/3c293c7b5708f2c551433057af517710/rf.o 
```

**Comment by daurnimator** [comment]
> Using `10.16` here is meaningless; but using `11.1.0` is appropriate.
> 
> ```
> ld64.lld -error-limit 0 -demangle -dynamic -arch x86_64 -macosx_version_min 10.16.0 -sdk_version 10.16.0 -pie -o rf zig-cache/o/3c293c7b5708f2c551433057af517710/rf.o 
> ```

Does this not work? What is the error message?

**Comment by mikdusan** [comment]
> Does this not work? What is the error message?

It immediately fails for an **indirect** reason -- some .zig code decides on passing `-syslibroot` only when `11.x` is detected and thus fails on my `11.x` system if using downloaded binaries, or using a binary I build myself against 10.15.* SDK and lower. (as an aside, I believe that I can come up with a better solution to the `-syslibroot` workaround from #6932,  more on that later).

#### fail (on Big Sur w/ zig built against 10.15.6 SDK)
```
ld64.lld -error-limit 0 -demangle -dynamic -arch x86_64 -macosx_version_min 10.16.0 -sdk_version 10.16.0 -pie -o rf zig-cache/o/3c293c7b5708f2c551433057af517710/rf.o /Users/mike/.cache/zig/o/9d95847e6b6f2162984f0cfd73b0ebad/libcompiler_rt.a -lSystem
warning(link): unexpected LLD stderr:
ld64.lld: error: Unable to find library for -lSystem
```

#### success (on Big Sur w/ zig built against 11.1.0 SDK)
```
ld64.lld -error-limit 0 -demangle -dynamic -arch x86_64 -macosx_version_min 11.1.0 -sdk_version 11.1.0 -pie -o rf -syslibroot /Applications/Xcode_12.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk zig-cache/o/7b9913ea831855db1526b0d9357f998b/rf.o /Users/mike/.cache/zig/o/4271907eac86b92cae46d62a868a23d4/libcompiler_rt.a -lSystem
```

but the deeper issue is when building against an SDK for `10.16.0` what does that mean? We've had 11.0 and 11.1 SDKs now, with 11.2 incoming. `10.16` is being returned for Big Sur 11.0.0, 11.0.1, 11.1.0 and 11.2 beta . It's quite meaningless except for general test to say "oh we're in backwards compatibility context". The linker may interpret `10.16.0` to mean exactly `11.0.0` but what about `11.0.1` or `11.1.0` or higher? Why does this matter? Because just as Zig builds with full detected CPU feature set by default, it also builds (links) against SDK based on detected OS version. When will `-macosx_version_min 10.16.0` fail? As soon as code uses symbols incompatible with `11.0`.

**Comment by kubkon** [comment]
Pasting in more context on breaking changes in Big Sur: https://github.com/ziglang/zig/issues/6318

One of those changes was Apple's decision to no longer store a copy of the relevant system dylibs in the usual places including `/usr/lib/`. Instead, when linking, we are now obliged to specify the root to the system libs using `-syslibroot` with value read from `$(xcrun --show-sdk-path)`.

**Comment by mikdusan** [comment]
I was under the impression that:

- linking still works if `dlopen()` is used instead of going the route of `opendir()`
- I think even LLD does the `dlopen()` method
- LLD **can** find `Metal` binary via `dlopen()` **only** if using the `/` default path
- LLD **cannot** find `Metal` when given `-syslibroot` because `/` is dropped from search, and SDK frameworks only has `Metal.tbd` which LLD doesn't even bother to search for <-- this is where the HACK environ for linker saves the day
- the SDK headers are deceptively problematic as well; example: an SDK 11.1 framework is shipped with SDK 11.1 /usr/include headers. Trying to use SDK 11.1 framework with whatever version of headers are bundled with zig doesn't work, many warnings until error'ing out on not finding `libDER/` headers that are not present in SDK 10.15 .
- and the final nail in the coffin is I cannot find any way to tell clang via args to ignore all previous include directives (ie. forget everything) and so zig's bundled `libSystem` headers cannot be undone via command line options; all hope is not lost I have a few other *hail marys* to try regarding over-riding zig's bundled includes without touching zig source code

**Comment by mikdusan** [comment]
This is how to quickly test zig executable's detection of Big Sur version:

#### show detected native triple:
```sh
zig targets | grep triple
```

bad output: Big Sur version is backwards compatible as `10.16`:
```
"triple": "x86_64-macos.10.16...10.16-gnu",
```

good output: when Big Sur version is a bonafide `11.1`:
```
"triple": "x86_64-macos.11.1...11.1-gnu",
```
